File: Android/app/src/main/java/com/hhp227/datemate/data/repository/UserRepository.kt
package com.hhp227.datemate.data.repository

import android.app.Activity
import com.google.firebase.auth.FirebaseUser
import com.hhp227.datemate.common.Resource
import com.hhp227.datemate.common.asResource
import com.hhp227.datemate.data.datasource.UserLocalDataSource
import com.hhp227.datemate.data.datasource.UserRemoteDataSource
import com.hhp227.datemate.data.model.UserCache
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.onStart

class UserRepository private constructor(
    private val userRemoteDataSource: UserRemoteDataSource,
    private val userLocalDataSource: UserLocalDataSource
) {
    val remoteUserStateFlow = userRemoteDataSource.userStateFlow

    val localUserStateFlow = userLocalDataSource.userFlow

    val signInStateFlow: Flow<SignInState> = localUserStateFlow
        .map { userCache ->
            if (userCache != null) {
                SignInState.SignIn
            } else {
                SignInState.SignOut
            }
        }
        .onStart { emit(SignInState.Loading) }

    fun getSignInResultStream(email: String, password: String): Flow<Resource<FirebaseUser>> {
        return flow {
            try {
                val user = userRemoteDataSource.signIn(email, password)

                emit(Resource.Success(user))
            } catch (e: Exception) {
                emit(Resource.Error(e.message ?: "로그인 실패"))
            }
        }
            .onStart { emit(Resource.Loading()) }
    }

    fun getSignUpResultStream(email: String, password: String): Flow<Resource<FirebaseUser>> {
        return flow {
            try {
                val user = userRemoteDataSource.signUp(email, password)

                emit(Resource.Success(user))
            } catch (e: Exception) {
                emit(Resource.Error(e.message ?: "회원가입 실패"))
            }
        }
            .onStart { emit(Resource.Loading()) }
    }

    fun getSignOutResultStream(): Flow<Resource<Boolean>> {
        return flow {
            try {
                emit(Resource.Success(userRemoteDataSource.signOut()))
            } catch (e: Exception) {
                emit(Resource.Error(e.message ?: "로그아웃 실패"))
            }
        }
            .onStart { emit(Resource.Loading()) }
    }

    fun getPasswordResetResultStream(email: String): Flow<Resource<Boolean>> {
        return flow {
            try {
                emit(Resource.Success(userRemoteDataSource.sendPasswordResetEmail(email)))
            } catch (e: Exception) {
                emit(Resource.Error(e.message ?: ""))
            }
        }
            .onStart { emit(Resource.Loading()) }
    }

    fun createUserProfileResultStream(userId: String, email: String?): Flow<Resource<Boolean>> {
        return flow {
            try {
                emit(Resource.Success(userRemoteDataSource.createUserProfile(userId, email)))
            } catch (e: Exception) {
                emit(Resource.Error(e.message ?: "프로필 생성 실패"))
            }
        }
            .onStart { emit(Resource.Loading()) }
    }

    suspend fun storeUserProfile(userCache: UserCache?) {
        userLocalDataSource.storeUser(userCache)
    }

    fun updateUserProfile(name: String, imageUrl: List<String>): Flow<Resource<Boolean>> {
        return flow {
            try {
                val result = userRemoteDataSource.updateUserProfile(
                    name,
                    imageUrl.firstOrNull()
                )

                emit(Resource.Success(result))
            } catch (e: Exception) {
                emit(Resource.Error(e.message ?: "프로필 업데이트 실패"))
            }
        }
            .onStart { emit(Resource.Loading()) }
    }

    fun sendOtp(phoneNumber: String, activityProvider: () -> Activity): Flow<Resource<String>> {
        return userRemoteDataSource.sendOtp(phoneNumber, activityProvider).asResource()
    }

    fun verifyOtp(verificationId: String, code: String): Flow<Resource<Boolean>> {
        return flow {
            try {
                userRemoteDataSource.verifyOtp(verificationId, code)
                emit(Resource.Success(true))
            } catch (e: Exception) {
                emit(Resource.Error(e.message ?: "OTP 검증 실패"))
            }
        }
            .onStart { emit(Resource.Loading()) }
    }

    enum class SignInState {
        SignIn, SignOut, Loading
    }

    companion object {
        @Volatile private var instance: UserRepository? = null

        fun getInstance(
            userRemoteDataSource: UserRemoteDataSource,
            userLocalDataSource: UserLocalDataSource
        ) =
            instance ?: synchronized(this) {
                instance ?: UserRepository(userRemoteDataSource, userLocalDataSource).also { instance = it }
            }
    }
}

File: iOS/DateMate/Data/Repository/UserRepository.swift
//
//  UserRepository.swift
//  DateMate
//
//  Created by 홍희표 on 2021/12/01.
//

import Foundation
import FirebaseAuth
import Combine

class UserRepository {
    private let userRemoteDataSource: UserRemoteDataSource
    
    private let userLocalDataSource: UserLocalDataSource
    
    private let storageRepository: StorageRepository
    
    private var cancellables = Set<AnyCancellable>()
    
    @Published var signInState: SignInState = .loading
    
    var signInStatePublisher: AnyPublisher<SignInState, Never> {
        $signInState.eraseToAnyPublisher()
    }
    
    var remoteUserStatePublisher: AnyPublisher<User?, Never> {
        userRemoteDataSource.userStatePublisher
    }

    var localUserStatePublisher: AnyPublisher<UserCache?, Never> {
        userLocalDataSource.userPublisher
    }
    
    func getSignInResultStream(email: String, password: String) -> AnyPublisher<Resource<FirebaseAuth.User>, Never> {
        userRemoteDataSource.signIn(email: email, password: password).asResource()
    }
        
    func getSignUpResultStream(email: String, password: String) -> AnyPublisher<Resource<FirebaseAuth.User>, Never> {
        userRemoteDataSource.signUp(email: email, password: password).asResource()
    }

    func getSignOutResultStream() -> AnyPublisher<Resource<Bool>, Never> {
        userRemoteDataSource.signOut().asResource()
    }
        
    func getPasswordResetResultStream(email: String) -> AnyPublisher<Resource<Bool>, Never> {
        userRemoteDataSource.sendPasswordResetEmail(email: email).asResource()
    }
    
    func fetchUserProfile(userId: String) -> AnyPublisher<Resource<Profile?>, Never> {
        userRemoteDataSource.fetchUserProfile(userId: userId).asResource()
    }
    
    func createUserProfile(userId: String, email: String?) -> AnyPublisher<Resource<Bool>, Never> {
        userRemoteDataSource.createUserProfile(userId: userId, email: email).asResource()
    }

    func storeUserProfile(_ cache: UserCache?) {
        userLocalDataSource.storeUser(cache)
    }
    
    func updateUserProfile(
        images: [URL],
        name: String,
        gender: String,
        birthdayMillis: Int64,
        bio: String,
        job: String,
        concurrency: Int = 4,
        retryCount: Int = 1
    ) -> AnyPublisher<Resource<String>, Never> {
        userRemoteDataSource.userStatePublisher
            .compactMap { $0 }
            .flatMap(maxPublishers: .max(1)) { user in
                self.storageRepository.uploadAllImages(
                    imageUrls: images,
                    userId: user.uid,
                    concurrency: concurrency,
                    retryCount: retryCount
                )
                .flatMap { uploadedUrls in
                    self.userRemoteDataSource.updateUserProfile(
                        userId: user.uid,
                        name: name,
                        gender: gender,
                        birthdayMillis: birthdayMillis,
                        bio: bio,
                        job: job,
                        profileImageUrls: uploadedUrls
                    )
                }
            }
            .asResource()
    }
    
    func sendOtp(phoneNumber: String) -> AnyPublisher<Resource<String>, Never> {
        userRemoteDataSource.sendOtp(phoneNumber: phoneNumber).asResource()
    }

    func verifyOtp(verificationId: String, code: String) -> AnyPublisher<Resource<Void>, Never> {
        userRemoteDataSource.verifyOtp(verificationId: verificationId, code: code).asResource()
    }
    
    init(
        _ userRemoteDataSource: UserRemoteDataSource,
        _ userLocalDataSource: UserLocalDataSource,
        _ storageRepository: StorageRepository
    ) {
        self.userRemoteDataSource = userRemoteDataSource
        self.userLocalDataSource = userLocalDataSource
        self.storageRepository = storageRepository
        
        userLocalDataSource.userPublisher
            .map { $0 != nil ? SignInState.signIn : SignInState.signOut }
            .prepend(.loading)
            .receive(on: DispatchQueue.main)
            .assign(to: \.signInState, on: self)
            .store(in: &cancellables)
    }
    
    private static var instance: UserRepository? = nil
    
    static func getInstance(
        userRemoteDataSource: UserRemoteDataSource,
        userLocalDataSource: UserLocalDataSource,
        storageRepository: StorageRepository
    ) -> UserRepository {
        if let instance = self.instance {
            return instance
        } else {
            let userRepository = UserRepository(
                userRemoteDataSource,
                userLocalDataSource,
                storageRepository
            )
            self.instance = userRepository
            return userRepository
        }
    }
}

enum SignInState {
    case signIn
    case signOut
    case loading
}
